Fix KDC DB propagation when larger than 4 GiB

jrisc opened the pull request:
When doing a full resync of the KDC database, `kprop` stores and sends the size of the dumped database on a 32 bits integer. As a consequence, any value bigger than 2<sup>32</sup>-1 (ie. >= 4GiB) wraps, resulting in a invalid smaller size.

I am not sure if this behavior could actually result in a truncated database being propagated without any error. It seems the deserialization will just fail if the invalid EOF happens at the middle of a line of the transmitted dumped DB file. But maybe it could happen in case invalid EOF occurs on a newline character. It does not looks like the database dump file contains any information about its length in the header.

This problem has been identified and reported to Red Hat by Keith Young from Morgan Stanley. He provided an initial fix I completed with tests, documentation, and some compatibility additions:

In order to remain compatible with previous versions of `kprop`, in case the size of the dumped KDC database is between 1 bit and 4GiB-1B, it is still encoded on 32 bits. The 0 value (which is impossible in practice) is used as an indicator that the size is stored on the 64 following bits, allowing a much larger size value. Hence a fixed `kprop` sending a > 4GiB dumped database to an older version of `kpropd` will see its size as 0, resulting in a header deserialization error (as mentioned in the additions to the documentation).

With this implementation of the compatibility approach described above, `kpropd` would consider valid an encoding where the first 32 bits integer is 0, but the next 64 bits one has a value smaller or equal to 4GiB. This should never happen given how encoding is implemented on the `kprop` side. Maybe such case should raise an error?

I mention the 4GiB limitation in `kprop` and `kpropd` manuals. Maybe this is not necessary? Please tell me what you think.

Also I wrote the documentation additions as if this fix was to be included into 1.20.0, but if I am being too optimistic, just let me know.

greghudson commented:
Apologies for the delay on this; I've been working on a larger pull request.

We don't normally document limitations or bugs in previous releases, and we don't normally create unit test programs for very simple functions like encode_database_size() and decode_database_size().

I will likely want to make adjustments to the encode and contracts for better length checking, but the general approach seems okay.

jrisc commented:
I removed documentation and tests.

I also modified `decode_database_size()` to be more restrictive and rely on `krb5_data.length` rather than the value of the first 4 bytes to infer the DB size type.

I used `KRB5KRB_ERR_GENERIC` as error. Would there be a more relevant one for this case?

greghudson commented:
I pushed some editorial changes.  I also added a functional test case to t_kprop.py, but disabled it unless an environment variable is set.  This is a first for the test suite, but it's not reasonable to produce, copy, and load a >4GB dump file during "make check" and there's no other way to verify that the code path works.  (The erstwhile unit tests didn't verify that the encoder and decoder were called correctly.)

Apologies again for the slow review.  Pending objections, this should be approximately ready to merge.

jrisc commented on src/kprop/kprop_util.c:
```diff
@@ -92,3 +92,46 @@ sn2princ_realm(krb5_context context, const char *hostname, const char *sname,
+    size = load_32_be(buf->data);
+    if (buf->length == 12) {
+        /* A 12 bytes size buffer must have the first four bytes zeroed. */
+        if (load_32_be(buf->data) != 0)
```
This call seems redundant with `size = load_32_be(buf->data)` at line 117.

jrisc commented on src/kprop/kprop_util.c:
```diff
@@ -92,3 +92,46 @@ sn2princ_realm(krb5_context context, const char *hostname, const char *sname,
+        if (size > 0 && size <= UINT32_MAX)
+            return KRB5KRB_ERR_GENERIC;
+    } else if (buf->length == 4) {
+        size = load_32_be(buf->data);
```
This case is supposed to be handled by line 117 too. If you prefer to have one block per case in the if/else/if statement, it's probably better to remove the line 117 and have each case block do the full decoding and/or error raising process.

greghudson replied:
Yeah, I meant to remove line 117 in the edits and apparently didn't.

jrisc reviewed (commented):
Just a remark about the changes in the `decode_database_size()` function
